python/semantic_kernel/connectors/memory/sql_server.py

```diff
@@ -0,0 +1,1028 @@
+    # Set the Matched clause
+    command.query.append("WHEN MATCHED THEN\n")
+    command.query.append("UPDATE SET ")
+    command.query.append_list(
+        [f"t.{field.name} = s.{field.name}" for field in chain(data_fields, vector_fields)], suffix="\n"
+    )
+    # Set the Not Matched clause
+    command.query.append("WHEN NOT MATCHED THEN\n")
+    with command.query.in_parenthesis("INSERT", " "):
+        _add_field_names(command.query, key_field, data_fields, vector_fields)
+    # add the closing parenthesis
+    with command.query.in_parenthesis("VALUES", " \n"):
+        _add_field_names(command.query, key_field, data_fields, vector_fields, table_identifier="s")
+    # add the closing parenthesis
+    command.query.append(f"OUTPUT inserted.{key_field.name} INTO @UpsertedKeys (KeyColumn);\n")
+    command.query.append("SELECT KeyColumn FROM @UpsertedKeys;\n")
+    return command
+
+
+def _build_select_query(
+    schema: str,
+    table: str,
+    key_field: VectorStoreRecordKeyField,
+    data_fields: list[VectorStoreRecordDataField],
+    vector_fields: list[VectorStoreRecordVectorField] | None,
+    keys: Sequence[Any],
+) -> SqlCommand:
+    """Build the SELECT query based on the data model."""
+    command = SqlCommand()
+    # start the SELECT statement
+    command.query.append("SELECT\n")
+    # add the data and vector fields
+    _add_field_names(command.query, key_field, data_fields, vector_fields)
+    # add the FROM clause
+    command.query.append_table_name(schema, table, prefix=" FROM", newline=True)
+    # add the WHERE clause
+    if keys:
+        command.query.append(f"WHERE {key_field.name} IN\n")
+        with command.query.in_parenthesis():
+            # add the keys
+            command.query.append_list(["?"] * len(keys))
+            command.add_parameters([_cast_value(key) for key in keys])
+    command.query.append(";")
+    return command
+
+
+def _build_delete_query(
+    schema: str,
+    table: str,
+    key_field: VectorStoreRecordKeyField,
+    keys: Sequence[Any],
+) -> SqlCommand:
+    """Build the DELETE query based on the data model."""
+    command = SqlCommand("DELETE FROM")
+    # start the DELETE statement
+    command.query.append_table_name(schema, table)
+    # add the WHERE clause
+    command.query.append(f"WHERE [{key_field.name}] IN")
+    with command.query.in_parenthesis():
+        # add the keys
+        command.query.append_list(["?"] * len(keys))
+        command.add_parameters([_cast_value(key) for key in keys])
+    command.query.append(";")
+    return command
+
+
+def _build_filter(filters: VectorSearchFilter) -> str | None:
+    """Build the filter query based on the data model."""
+    if not filters.filters:
+        return None
+    query_builder = QueryBuilder("WHERE ")
+    for filter in filters.filters:
+        match filter:
+            case EqualTo():
+                query_builder.append(f"[{filter.field_name}] = '{_cast_value(filter.value)}' AND\n")
```

Comment:
Shouldn't we use `command.add_parameter(...)` here as well? It feels like we could be concatenating untrusted input directly into the query string.

Reply:
this is a very good catch, thanks @moonbox3, fixed now.